Line.vala

```diff
@@ -74,40 +76,76 @@ class Line {
+			set_background(attributes, start, end, style.background);
+		}
+		if (style.weight != null) {
+			set_weight(attributes, start, end, style.weight);
+		}
+		if (style.italic) {
+			set_italic(attributes, start, end);
+		}
+	}
+
 	public void set_styles(Json.Array styles) {
-		var attributes = new Pango.AttrList();
+		StyleMap style_map = StyleMap.get_instance();
+		selection_ranges = {};
+		text_attributes = new Pango.AttrList();
 		uint offset = 0;
 		for (int i = 0; i < styles.get_length(); i += 3) {
 			uint start = offset + (uint)styles.get_int_element(i);
 			uint end = start + (uint)styles.get_int_element(i+1);
+			offset = end;
+
 			int style_id = (int)styles.get_int_element(i+2);
-			Style style = StyleMap.get_instance().get_style(style_id);
-			if (style.foreground != null) {
-				set_foreground(attributes, start, end, style.foreground);
-			}
-			if (style.background != null) {
-				set_background(attributes, start, end, style.background);
-			}
-			if (style.weight != null) {
-				set_weight(attributes, start, end, style.weight);
-			}
-			if (style.italic) {
-				set_italic(attributes, start, end);
+
+			// add selections to selection array to defer conversion to Pango.AttrList until draw-time
+			if (style_id == 0) {
+				selection_ranges += start;
+				selection_ranges += end;
+				continue;
 			}
-			offset = end;
+
+			Style style = style_map.get_style(style_id);
+			apply_style_to_range(text_attributes, style, start, end);
 		}
-		layout.set_attributes(attributes);
 	}
 
-	public void draw(Cairo.Context cr, double x, double y, double width, double ascent, double line_height, bool draw_cursors) {
+	public void draw(Cairo.Context cr, Gtk.StyleContext style_ctx, double x, double y, double width, double ascent, double line_height, bool draw_cursors) {
+		var widget_state = style_ctx.get_state();
+		var selected_state = widget_state | Gtk.StateFlags.SELECTED;
+		// highlight line if it contains any cursors
 		if (cursors.length > 0) {
-			Gdk.cairo_set_source_rgba(cr, Utilities.convert_color(0xfff5f5f5u));
+			var color = style_ctx.get_background_color(selected_state);
+			color.alpha /= 2;
+			Gdk.cairo_set_source_rgba(cr, color);
 			cr.rectangle(0, y, width, line_height);
 			cr.fill();
 		}
-		Gdk.cairo_set_source_rgba(cr, Utilities.convert_color(0xff323232u));
 		cr.move_to(x, y + ascent);
+
+		// create a style for the selection
+		Style selected_style = Style() {
+			background = style_ctx.get_background_color(selected_state),
+			foreground = style_ctx.get_color(selected_state)
+		};
+		// merge selection with existing text attributes
+		var new_attributes = text_attributes.copy();
+		for (var i = 0; i < selection_ranges.length; i+=2) {
+			var start = selection_ranges[i];
+			var end = selection_ranges[i+1];
+			apply_style_to_range(new_attributes, selected_style, start, end);
+		}
+		// draw text
+		Gdk.cairo_set_source_rgba(cr, style_ctx.get_color(widget_state));
+		layout.set_attributes(new_attributes);
```

Comment:
Which work? Style creation is just an object allocation and a couple writes; the attribute copy is O(n) in length of the attribute list (as is drawing) and unavoidable in general, but we could cache to avoid redoing if neither state nor style has changed since last draw of this line; we could also avoid that copy if there are no selections on this line.
The iteration over selection_ranges shouldn't be a large cost, but it is doing O(n²) work in the number of styles *just like the old code* which this commit moves from `set_styles` to `apply_style_to_range`.

`layout.set_attributes` is an O(1) refcount bump and pointer assignment in pango.

I don't think asymptotic efficiency has actually changed here; if there are benchmarks I can run to quantify this intuition it would be great. Informally, I played with large files and found pathological cases like #19 prior to these changes as part of understanding the time complexity of the current code and making sure I wasn't accidentally hurting performance. I don't notice any worsening of #19 or normal editor usage with these changes.